### PR TITLE
Addresses #13216: use of type classes in the return clause of a match.

### LINKED ISSUE #13216: Reduced support for type classes in the inference of the return type of a match
#### Description of the problem

While porting the containers contrib, I found this change of behavior which (a priori??) is not deliberate:
```
Class A.
Declare Instance a:A.
Inductive T `(A) := C.
Fail Definition f x := match x with C _ => 0 end.
(* Cannot infer the 1st argument of the inductive type (T) of this term (no type
class instance found). *)
```

#### Coq Version

Since commit fb1c2a017e (in version 8.10).

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

This was deactivated in fb1c2a017e, probably without strong will to do so, but it seems useful (e.g. in contribs containers).

Fixes / closes #13216

- [X] Added / updated test-suite
- [x] Entry added in the changelog

### CODE REVIEW

**Comment by ppedrot** [comment]
Just out of curiosity I think this would deserve a bench.

File: doc/changelog/02-specification-language/13217-master+fix13216-typeclass-for-match-return-clause.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- **Fixed:**
+  Allow use of type classes inference for the return predicate of a :n:`match`
+  (was deactivated in versions 8.10 to 8.12, `#13217 <https://github.com/coq/coq/pull/13217>`_,
+  fixes `#13216 <https://github.com/coq/coq/issues/13216>`_,
+  by Hugo Herbelin).
```
Comment (ppedrot) [inline doc/changelog/02-specification-language/13217-master+fix13216-typeclass-for-match-return-clause.rst:5]: You should squash the two change commits.

**Review by ppedrot — APPROVED** [review]
Modulo bench and squash, the code changes are fine.

**Comment by herbelin** [comment]
For a bench, I clicked on the bench job of the gitlab pipeline, leading to https://gitlab.com/coq/coq/-/jobs/832712529. Is that the correct (new) way to do? If yes, it is ±0.06%. Isn't it a low (as a noise)? And a -7% memory for HoTT which looks strange.

But anyway, I don't expect an impact. The situation where the return predicate is a class, and its arguments are not already inferred by typing are probably rare. As a matter of fact, the 2018 change of semantics has been mostly unnoticed, except in one user contribution, so I see no reason that the CI would use it more now.

**Comment by ppedrot** [comment]
Bench:
```
┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │    mem faults     │
│                             │                         │                                             │                                             │                               │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-coqprime │   79.18   79.87 -0.86 % │      218097659795      218742616479 -0.29 % │      305898326660      305955976582 -0.02 % │     763324     761384 +0.25 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-color │  512.65  515.00 -0.46 % │     1429884001488     1433217935346 -0.23 % │     1793452772698     1793411063890 +0.00 % │    1133312    1133296 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fiat-core │  102.14  102.59 -0.44 % │      291700123272      291931024747 -0.08 % │      386296863227      386025885924 +0.07 % │     479292     479412 -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-verdi │  102.20  102.62 -0.41 % │      281764381859      282305523501 -0.19 % │      395673715499      395518287963 +0.04 % │     547304     547444 -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  221.60  222.39 -0.36 % │      617439193058      618211036475 -0.12 % │      935770357244      935654336905 +0.01 % │     667480     666292 +0.18 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-math-classes │  211.92  212.59 -0.32 % │      591052227665      591853189173 -0.14 % │      810174623629      810134011517 +0.01 % │     507236     507140 +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  632.35  634.22 -0.29 % │     1772263539703     1775036144976 -0.16 % │     2666495752302     2666501526670 -0.00 % │    3651028    3651072 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1441.24 1444.72 -0.24 % │     4020364000132     4022087345453 -0.04 % │     6084169529954     6084115950065 +0.00 % │    1026244    1027976 -0.17 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-flocq │  156.78  157.01 -0.15 % │      435136545665      435928909825 -0.18 % │      540713046984      540613732591 +0.02 % │     821816     822120 -0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-corn │ 1624.80 1627.01 -0.14 % │     4526092362295     4529952180428 -0.09 % │     6923360649169     6923425242472 -0.00 % │     838360     839940 -0.19 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  139.78  139.96 -0.13 % │      388703152413      388960750521 -0.07 % │      497633290689      497602820788 +0.01 % │     545556     545484 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-bignums │   56.49   56.56 -0.12 % │      156549457349      156424333504 +0.08 % │      210617966080      210612035574 +0.00 % │     469768     469688 +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  169.48  169.66 -0.11 % │      472154082775      472024937137 +0.03 % │      664971922522      664956457853 +0.00 % │     668056     675712 -1.13 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  154.98  155.13 -0.10 % │      430867030564      431632357735 -0.18 % │      595455641647      595512573574 -0.01 % │     709692     709220 +0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-unimath │ 3896.52 3900.12 -0.09 % │    10841964835686    10853188131679 -0.10 % │    20638455893098    20635950058655 +0.01 % │    1129372    1129236 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │  790.58  791.22 -0.08 % │     2201048614147     2200990698624 +0.00 % │     3410428217240     3410932580074 -0.01 % │     970524     970412 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2562.31 2564.30 -0.08 % │     7142422784491     7148661358130 -0.09 % │    11782351889529    11782130989539 +0.00 % │     764596     764572 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-rewriter │  719.06  719.55 -0.07 % │     1995123122262     1995887946873 -0.04 % │     3041864395076     3041651421699 +0.01 % │     974968     975192 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1046.05 1046.52 -0.04 % │     2915612299766     2914705172920 +0.03 % │     4622853650080     4622636796888 +0.00 % │     975076     975088 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1242.26 1242.45 -0.02 % │     3457271676450     3457382586930 -0.00 % │     4916364721102     4916013963127 +0.01 % │    1114008    1114184 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  222.35  222.18 +0.08 % │      615537728725      614869314674 +0.11 % │     1019024370314     1018537288731 +0.05 % │    1086724    1086836 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-engine-bench-lite │  367.10  366.69 +0.11 % │     1032492935173     1031698712004 +0.08 % │     1739985608161     1741094577909 -0.06 % │    3986444    3986728 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   47.42   47.36 +0.13 % │      131251204797      131126640086 +0.09 % │      180321079392      180355781160 -0.02 % │     480952     481004 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-compcert │  642.95  642.12 +0.13 % │     1791303347926     1788828396422 +0.14 % │     2466824428534     2466510627272 +0.01 % │    1047264    1047112 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-fiat-crypto │ 5068.92 5061.71 +0.14 % │    14080914031223    14063903143948 +0.12 % │    24229907208836    24223463728697 +0.03 % │    2273372    2268452 +0.22 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1223.48 1221.64 +0.15 % │     3407694594900     3402024060471 +0.17 % │     4821900894838     4822423320827 -0.01 % │     902020     902104 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                         coq │  517.99  516.90 +0.21 % │     1450144456351     1449630321700 +0.04 % │     1945921338882     1945949577544 -0.00 % │     562956     562920 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   50.09   49.94 +0.30 % │      137061190869      137057147850 +0.00 % │      173782233746      173760620118 +0.01 % │     513920     514012 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-hott │  323.36  322.36 +0.31 % │      877029991097      875710899708 +0.15 % │     1421332633895     1421171164767 +0.01 % │     679076     734724 -7.57 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-coqutil │   60.33   59.65 +1.14 % │      167439966501      165886879797 +0.94 % │      206756976129      206717302321 +0.02 % │     524764     524592 +0.03 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴───────────────────┘
```
Within noise range, let's merge.

**Pull request merged**